src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchWriter.cs

```diff
@@ -0,0 +1,651 @@
+#endif
+
+        /// <summary>
+        /// Starts a new batch - implementation of the actual functionality.
+        /// </summary>
+        protected override void WriteStartBatchImplementation()
+        {
+            WriteBatchEnvelope();
+            this.SetState(BatchWriterState.BatchStarted);
+        }
+
+        /// <summary>
+        /// Given an enumerable of dependsOn ids containing request ids and group ids, convert the group ids
+        /// into associated request ids.
+        /// </summary>
+        /// <param name="dependsOnIds">The dependsOn ids specifying current request's prerequisites.</param>
+        /// <returns>An enumerable consists of request ids.</returns>
+        protected override IEnumerable<string> GetDependsOnRequestIds(IEnumerable<string> dependsOnIds)
+        {
+            List<string> dependsOnRequestIds = new List<string>();
+            foreach (string id in dependsOnIds)
+            {
+                if (this.atomicityGroupIdToRequestId.ContainsKey(id))
+                {
+                    dependsOnRequestIds.AddRange(this.atomicityGroupIdToRequestId[id]);
+                }
+                else
+                {
+                    dependsOnRequestIds.Add(id);
+                }
+            }
+
+            return dependsOnRequestIds;
+        }
+
+        /// <summary>
+        /// Ends a batch - implementation of the actual functionality.
+        /// </summary>
+        protected override void WriteEndBatchImplementation()
+        {
+            // write pending message data (headers, response line) for a previously unclosed message/request
+            this.WritePendingMessageData(true);
+
+            this.SetState(BatchWriterState.BatchCompleted);
+
+            // Close the messages array
+            jsonWriter.EndArrayScope();
+
+            // Close the top level scope
+            jsonWriter.EndObjectScope();
+        }
+
+        /// <summary>
+        /// Starts a new changeset - implementation of the actual functionality.
+        /// </summary>
+        /// <param name="groupId">
+        /// The atomic group id of the changeset to start.
+        /// If it is null for Json batch, an GUID will be generated and used as the atomic group id.</param>
+        protected override void WriteStartChangesetImplementation(string groupId)
+        {
+            Debug.Assert(this.atomicityGroupId == null, "this.atomicityGroupId == null");
+
+            // write pending message data (headers, response line) for a previously unclosed message/request
+            this.WritePendingMessageData(true);
+
+            // important to do this first since it will set up the change set boundary.
+            this.SetState(BatchWriterState.ChangesetStarted);
+
+            // When the changeset start is created for request, the <paramref name="groupId"/> could be null
+            // and we will create an GUID here.
+            this.atomicityGroupId = groupId ?? Guid.NewGuid().ToString();

```

Comment:
create the groupid in the parameterless WriteStartChangeset() of the base class; assert in this method that it is non-null.